examples/scripts/spme_max_energy.py

```diff
@@ -0,0 +1,175 @@
+import pybop
+import pybamm
+import numpy as np
+import warnings
+
+# A design optimisation example loosely based on work by L.D. Couto
+# available at https://doi.org/10.1016/j.energy.2022.125966.
+
+# The target is to maximise the gravimetric energy density over a
+# range of possible design parameter values, including for example:
+# cross-sectional area = height x width (only need change one)
+# electrode widths, particle radii, volume fractions and
+# separator width.
+
+# Define parameter set and model
+parameter_set = pybop.ParameterSet.pybamm("Chen2020")
+model = pybop.lithium_ion.SPMe(parameter_set=parameter_set)
+
+# Fitting parameters
+parameters = [
+    pybop.Parameter(
+        "Positive electrode thickness [m]",
+        prior=pybop.Gaussian(7.56e-05, 0.05e-05),
+        bounds=[7e-05, 10e-05],
+    ),
+    pybop.Parameter(
+        "Positive particle radius [m]",
+        prior=pybop.Gaussian(5.22e-06, 0.05e-06),
+        bounds=[3e-06, 9e-06],
+    ),
+]
+
+
+# Define functions
+def nominal_capacity(x, model):
+    """
+    Update the nominal capacity based on the theoretical energy density and an
+    average voltage.
+    """
+    inputs = {
+        key: x[i] for i, key in enumerate([param.name for param in model.parameters])
+    }
+    model._parameter_set.update(inputs)
+
+    theoretical_energy = model._electrode_soh.calculate_theoretical_energy(
+        model._parameter_set
+    )
+    average_voltage = (
```

Comment:
The nominal capacity is required to scale the C-rate used in the experiment, so it must be computed before the simulation. However, it is only a scaling factor on the current so, as long as the same `average_voltage` is used each time, I believe the energy density will be maximised.

Reply:
Understood. This implementation would result in an optimal energy density for a given applied current (scaled by the calculated nom. capacity), but this applied current wouldn't be the correct C-rate, right? I think we can get the correct mean voltage via the OCV function though. 

Something like this `model._parameter_set["Positive electrode OCP [V]"](mean_sto)` where the`mean_sto` variable can be calculated by taking the mean of`model._esoh_solver.get_min_max_stoichiometries()`. More information [here](https://github.com/pybamm-team/PyBaMM/blob/903323efca7fef8f630924d114ce70dcf8a92363/pybamm/models/full_battery_models/lithium_ion/electrode_soh.py#L891).

Reply:
As far as I understand, there's no feasible way to compute the "correct C-rate" - this is a challenge that comes with using C-rate instead of a defined current and one that we will probably have to address more explicitly in user guidance in the future. In practice, the measured charge/discharge capacity depends on the electrode balancing as well as the overpotential dynamics. So the question becomes: how accurate does the C-rate need to be for the optimisation results to hold and by how much do we want to prioritise accuracy over speed?

The approach you suggest will certainly give a more accurate "average_voltage" but it will also increase the computational load, which is already high due to the re-building. It may also disguise the fact that we need to use an approximation here to obtain a C-rate pre-experiment. We could calculate the realistic mean once at the beginning, using the starting parameter set and inputs, but this would change the cost values from run-to-run so reproducibility could be an issue.

A fixed, easy-to-compute value avoids these issues and, while a particular voltage use-case may be more helpful, I think an average of the voltage limits is a reasonable approximation for now.